### PR TITLE
Api 82/api independent rate limit

### PR DESCRIPTION
**Pull request by dempe**
@ThorArakis 

Schema looks like
```
cqlsh:kong> select * from sm_ratelimiting_metrics ;

 account_id                           | user_id                              | period_date              | period | value
--------------------------------------+--------------------------------------+--------------------------+--------+-------
 7f5ee815-9fd3-42e6-ba4c-32dd8b58d4cb | 081d8171-f307-4d2d-963f-0711bd3612ba | 2016-05-25 21:40:41+0000 | second |     1
 7f5ee815-9fd3-42e6-ba4c-32dd8b58d4cb | 081d8171-f307-4d2d-963f-0711bd3612ba | 2016-05-25 21:33:00+0000 | minute |     1
 7f5ee815-9fd3-42e6-ba4c-32dd8b58d4cb | 081d8171-f307-4d2d-963f-0711bd3612ba | 2016-05-25 21:31:04+0000 | second |     1
 7f5ee815-9fd3-42e6-ba4c-32dd8b58d4cb | 94f3301e-f95d-4980-aa2a-f4fd0b20c147 | 2016-05-01 00:00:00+0000 |  month |     4
 7f5ee815-9fd3-42e6-ba4c-32dd8b58d4cb | 081d8171-f307-4d2d-963f-0711bd3612ba | 2016-05-01 00:00:00+0000 |  month |     4
```

https://simplymeasured.jira.com/browse/API-79 Will need to be in before we merge this though.

### CODE REVIEW

**Comment by dempe** [comment]
Closing as does not meet the following requirements:

- If it’s a viper jwt (no rate limiting built in):
  - there should be a global rate limit per user-id 
  - account_id will not be present in the jwt
  - ratelimit will not be present in the jwt
- If it’s a direct api JWT:
  - there should be a monthly rate limit base ​_only_​ on account_id
  - the rate limit and account_id should be in the jwt
  - the monthly ratelimit is across all users

File: kong/plugins/sm-rate-limiting/handler.lua [hunk 1]
```diff
@@ -42,15 +46,41 @@ local function get_user_id()
...
+
+local function get_account_id()
+    local headers = req_get_headers()
+    local authHeader = headers["Authorization"]
+
+    if not authHeader then
+        return responses.send_HTTP_BAD_REQUEST("JWT authorization must be enabled to use this plugin.")
+    end
+
+    local jwt, err = jwt_decoder:new(authHeader:gsub("[B,b]earer ", ""))
```
Comment (ThorArakis) [inline kong/plugins/sm-rate-limiting/handler.lua]: It looks like you're decoding the jwt multiple times. I don't think it's a huge overhead, but it's still unnecessary. You should move the jwt decoding to it's own method and then pass the obj into get_account and get_user and eventually get_ratelimits

**Comment by ThorArakis** [comment]
Also as this is our most complicated plugin behavior so far, we need to add tests. Check out https://getkong.org/docs/0.5.x/plugin-development/tests/

**Comment by ThorArakis** [comment]
This is definitely a good start so far though, nice progress!

File: kong/plugins/sm-rate-limiting/handler.lua (hunk 1 above)
Comment (dempe) [inline kong/plugins/sm-rate-limiting/handler.lua]: Derp.  👍

**Pull request closed without merging**